src/axom/inlet/examples/mfem_coefficient.cpp

```diff
@@ -0,0 +1,190 @@
+
+#include "axom/inlet.hpp"
+
+#include <iostream>
+#include <unordered_map>
+#include "CLI11/CLI11.hpp"
+#include "axom/slic/core/UnitTestLogger.hpp"
+#include "mfem.hpp"
+
+using axom::inlet::Inlet;
+using axom::inlet::LuaReader;
+using axom::sidre::DataStore;
+
+namespace inlet = axom::inlet;
+
+// This example relies on mfem's support for std::function-based coefficients
+#if MFEM_VERSION_MAJOR >= 4 && MFEM_VERSION_MINOR >= 2
+  #define MFEM_STDFUNCTION_COEF
+#endif
+
+#ifdef MFEM_STDFUNCTION_COEF
+
+struct BoundaryCondition
+{
+  std::unordered_map<int, int> attrs;
+  // These act as a sort of variant without C++17
+  std::unique_ptr<mfem::Coefficient> coef;
+  std::unique_ptr<mfem::VectorCoefficient> vec_coef;
+
+  static void defineSchema(inlet::Table& schema)
+  {
+    schema.addIntArray("attrs", "List of boundary attributes");
+    // Inlet does not support sum types, so options are added to the schema
+    // for vector/scalar coefficients and for time-dependent versions of each
+    // Supported function parameter/return types are Double and Vec3D
+    schema.addFunction("vec_coef",
+                       inlet::FunctionType::Vec3D,    // Return type
+                       {inlet::FunctionType::Vec3D},  // Argument type
+                       "The function representing the BC coefficient");
+
+    schema.addFunction("coef",
+                       inlet::FunctionType::Double,
+                       {inlet::FunctionType::Vec3D},
+                       "The function representing the BC coefficient");
+
+    schema.addFunction("vec_coef_t",
+                       inlet::FunctionType::Vec3D,
+                       {inlet::FunctionType::Vec3D,
+                        inlet::FunctionType::Double},  // Multiple argument types
+                       "The function representing the BC coefficient");
+
+    schema.addFunction("coef_t",
+                       inlet::FunctionType::Double,
+                       {inlet::FunctionType::Vec3D, inlet::FunctionType::Double},
+                       "The function representing the BC coefficient");
+  }
+};
+
+/**
+ * Example Lua definition:
+ * \code{.lua}
+ * bc = {
+ *   attrs = {
+ *      3, 4, 6, 9
+ *   }
+ *   coef = function (x, y, z)
+ *     return x * 0.12
+ *   end
+ * }
+ * -- or, for vector coefficients:
+ * [8] = {
+ *   attrs = { [4] = 14, [8] = 62, [6] = 11},
+ *   vec_coef = function (x, y, z)
+ *     scale = 0.12
+ *     return x * scale, y * scale, z * scale
+ *   end
+ * }
+ * \endcode
+ */
+template <>
+struct FromInlet<BoundaryCondition>
+{
+  BoundaryCondition operator()(const inlet::Table& base)
+  {
+    BoundaryCondition bc;
+    bc.attrs = base["attrs"];
+    if(base.contains("vec_coef"))
+    {
+      // We assume the dimension is 3 - if the mfem::Vector is 2D,
+      // the z-component will be set to zero due to how primal::Vector's
+      // (T*, size) constructor works
+      bc.vec_coef = std::make_unique<mfem::VectorFunctionCoefficient>(
+        3,
```

Comment:
I believe the `mfem::VectorFunctionCoefficient` must be set to `dim=2` if the mesh is 2D. Is there a way to get around this hard coded variable?

Reply:
I think we would have to use a similar approach as we do for the mesh in serac, namely, reading in all relevant information from the input file and then constructing the actual objects using the data from Inlet plus the instance-specific data (in this case, mesh dimension.)